ESQL: Correct out-of-range filter pushdowns

alex-spies opened the pull request:
Fix pushed down filters for binary comparisons that would implicitly compare a byte/short/int/long with an out of range value, like `WHERE some_int_field < 1E300`. Queries containing such comparisons lead to exceptions thrown by ElasticSearch. Instead, trivialize the query, either to a query that matches all single values or no values depending on the comparison.

This does not cover the same problem for half_float/scaled_float/float,
see https://github.com/elastic/elasticsearch/issues/100130.

Closes https://github.com/elastic/elasticsearch/issues/99960.

alex-spies committed:
Fix unsupported filter pushdown into Lucene

Prevent pushing down filters for binary comparisons that would
implicitly compare a byte/short/int with an out of range value. This
leads to exceptions thrown by Lucene - instead, evaluate the filter in
ESQL only.

This does not cover the same problem for half_float/scaled_float/float,
see https://github.com/elastic/elasticsearch/issues/100130.

Closes https://github.com/elastic/elasticsearch/issues/99960.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LocalPhysicalPlanOptimizer.java:
```diff
@@ -194,6 +196,8 @@ private static Set<Attribute> missingAttributes(PhysicalPlan p) {
     }
 
     public static class PushFiltersToSource extends OptimizerRule<FilterExec> {
+        private static int HALF_FLOAT_MAX = 65504;
```
I did not find this constant defined in the code base - anyone know if it is defined somewhere, after all?

elasticsearchmachine commented:
Pinging @elastic/es-ql (Team:QL)

elasticsearchmachine commented:
Hi @alex-spies, I've created a changelog YAML for you.

alex-spies committed:
Update docs/changelog/99961.yaml

costin reviewed (commented):
I'm not sure this is the correct solution in the long term - emp_no is an integer so the comparison fails because the value that it is compared against is out of range.
Doing the comparison in ESQL happens because we're widening the types but we sacrifice performance.

The issue boils down to whether or not the widening should take place or not - inside filters I would argue this should translate into a NoMatch since the data is out of range.
In other words the type should be narrowed and, in case of failure, resolve the filter to false.

alex-spies commented:
> I'm not sure this is the correct solution in the long term - emp_no is an integer so the comparison fails because the value that it is compared against is out of range. Doing the comparison in ESQL happens because we're widening the types but we sacrifice performance.
> 
> The issue boils down to whether or not the widening should take place or not - inside filters I would argue this should translate into a NoMatch since the data is out of range. In other words the type should be narrowed and, in case of failure, resolve the filter to false.

Agree, long term we'd expect the filters to become trivial if we compare to out-of-range values. That should probably go into some other optimization rule, though, or into the filter logic itself.

My (short-term) goal with this PR is to not fail the queries since I think they are valid per ES|QL's semantics. This is not just about widening, but about allowing comparisons between different numerical types. `... where integer == 1.0` is a valid query, and it's unexpected that  `... where integer == 1000000000000` is not.

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into esql-fix-wrong-type-in-lucene-filter

alex-spies committed:
Spotless

alex-spies committed:
Make Checkstyle happy

elasticsearchmachine commented:
Pinging @elastic/es-analytics-geo (Team:Analytics)

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into esql-fix-wrong-type-in-lucene-filter

alex-spies committed:
Fix tests

alex-spies commented:
I updated the PR to make the CI happy again.

Next, I want to align this closer with  Costin's comment (out of range comparisons should be no-ops); no need to (re-) review until then.

alex-spies committed:
Minor cleanup

alex-spies committed:
Optimize away out-of-range comparisons

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into esql-fix-wrong-type-in-lucene-filter

alex-spies committed:
Improve javadoc

alex-spies commented:
Added an optimization rule that trivializes out of range comparisons altogether - so that `WHERE  some-int-field > (2147483648 + 1)` becomes `WHERE false` because the right hand side is already larger than any possible integer.

This is at least bordering on premature optimization since these comparisons may happen very rarely: they require a _literal_ that is out of range, so the user is kinda sending a non-sensical query in the first place.

However, in cases where this happens, preventing filter pushdowns altogether is not nice, either, and maintenance cost for this optimization rule is probably okay.

@costin , let me know if you agree with this approach - if so, this needs a small clean up + improved tests before being mergeable.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:
```diff
@@ -612,6 +622,65 @@ protected Expression maybeSimplifyNegatable(Expression e) {
 
     }
 
+    /**
+     * Look for numerical literals on the right hand side of a {@link BinaryComparison}s that are out of range for the left hand side data
+     * type and replace the comparison by a literal TRUE or FALSE. This rule should be placed after {@link LiteralsOnTheRight}.
+     */
+    private static class OutOfRangeBinaryComparison extends OptimizerRules.OptimizerExpressionRule<BinaryComparison> {
```
This is the added optimization rule.

costin commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LocalPhysicalPlanOptimizerTests.java:
```diff
@@ -152,12 +146,124 @@ public void init() {
+     * \_ExchangeExec[[],false]
+     *   \_EsQueryExec[test], query[][_doc{f}#19], limit[500], sort[] estimatedRowSize[4]
+     */
+    public void testTrivialFilterForUnsupportedImplicitCast() {
```
The tests are duplicated across the rest tests and here - move the beef of the tests here and add a couple on the rest layer.
Since this impacts Lucene, another approach is to create a simple index, test the generated query and then look at its output. However at that point your retesting the conversion and not your rule.

costin commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LocalPhysicalPlanOptimizerTests.java:
```diff
@@ -171,7 +277,7 @@ public void testCountAllWithEval() {
               from test | eval s = salary | rename s as sr | eval hidden_s = sr | rename emp_no as e | where e < 10050
             | stats c = count(*)
-            """);
+            """, basicMappingAnalyzer);
```
Refactoring tests should simplify them not make them more complicated. 
All these tests expect an analyzer and should not require a basic piece of infrastructure to be plugged in.
Potentially overload the plan method to accept a different mapping and create an analyzer from it.

costin commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/Literal.java:
```diff
@@ -32,6 +32,14 @@ public Literal(Source source, Object value, DataType dataType) {
         this.value = value;
     }
 
+    public static Literal trueWithSource(Source source) {
+        return new Literal(source, Boolean.TRUE, DataTypes.BOOLEAN);
+    }
+
+    public static Literal falseWithSource(Source source) {
+        return new Literal(source, Boolean.FALSE, DataTypes.BOOLEAN);
+    }
+
```
QL classes need to remain slim - these type of utility methods are more trouble than their worth.
There are already utilities that provide similar functionality - see my previous comments.

costin commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerUtils.java:
```diff
@@ -0,0 +1,93 @@
+
+package org.elasticsearch.xpack.ql.optimizer;
+
+public class OptimizerUtils {
```
The QL module is shared by multiple projects - unless there's a concrete need to have code in multiple places, work only in ESQL.
There's no other place where these methods are used so move them inside the optimizer rule. We can figure out in time if there's an opportunity for reuse.

alex-spies replied:
Yes, absolutely - already planned to move this into the esql project.

costin commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerUtils.java:
```diff
@@ -0,0 +1,93 @@
+    /**
+     * Check if an object  is a numerical value that fits into an integer.
+     */
+    public static boolean isInIntegerRange(Object num) {
```
All these classes implement Number so this can be simplified to 
Number.longValue() < > Integer.MIN/MAX
However this will likely not maintain the original range so widening to long and double is necessary (which is what the JVM does behind the scene).

This is a typical Java gotcha hence why you need to check this works as expected.

alex-spies replied:
Ended up using big decimals to deal with the widening in a simple way.

costin commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerUtils.java:
```diff
@@ -0,0 +1,93 @@
+    /**
+     * Check if a folded expression is a positive numerical value.
+     */
+    public static boolean isPositive(Object num) {
```
See Math/Integer/Long.signum.

alex-spies replied:
Ended up just using `Number.doubleValue` which should work for all types.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:
```diff
@@ -612,6 +622,65 @@ protected Expression maybeSimplifyNegatable(Expression e) {
+    private static class OutOfRangeBinaryComparison extends OptimizerRules.OptimizerExpressionRule<BinaryComparison> {
+
+        OutOfRangeBinaryComparison() {
+            super(TransformDirection.DOWN);
```
The direction should be up, bottoms-up since the constants are at the bottom.

alex-spies replied:
I initially thought this, too; however, semantically the direction doesn't matter for the outcome, and going down has the benefit that we avoid  traversing a whole subtree if we encounter an out of range literal.

E.g. transforming `(a + b*(c-d*(e-f)) < 1E307` is better done top-down because, if we know the data type on the left is int, we don't need to descend into the left hand side at all to know this comparison is always true.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:
```diff
@@ -612,6 +622,65 @@ protected Expression maybeSimplifyNegatable(Expression e) {
+        @Override
+        protected Expression rule(BinaryComparison bc) {
+            // TODO: add tests with null literals if they do not already exist
+            if (bc.left().dataType().isNumeric() == false || bc.right().dataType().isNumeric() == false) {
```
BinaryComparison accepts only numbers - the fact that it reached the optimizer means the plan has been already validated.
See BinaryComparisonSimplification above

alex-spies replied:
Fair enough - I wanted to be defensive here, but assuming that `BinaryComparison` only has numerical arguments makes sense.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:
```diff
@@ -612,6 +622,65 @@ protected Expression maybeSimplifyNegatable(Expression e) {
+                    return bc;
+                }
+
+                if (bc instanceof org.elasticsearch.xpack.ql.expression.predicate.operator.comparison.Equals || bc instanceof NullEquals) {
+                    return Literal.falseWithSource(bc.source());
+                }
+                if (bc instanceof NotEquals) {
+                    return Literal.trueWithSource(bc.source());
+                }
+
+                boolean isPositive = OptimizerUtils.isPositive(r);
+                if (bc instanceof GreaterThan || bc instanceof GreaterThanOrEqual) {
+                    return isPositive ? Literal.falseWithSource(bc.source()) : Literal.trueWithSource(bc.source());
+                }
+                if (bc instanceof LessThan || bc instanceof LessThanOrEqual) {
+                    return isPositive == false ? Literal.falseWithSource(bc.source()) : Literal.trueWithSource(bc.source());
+                }
```
In all cases the conditions create a true/false so do that and then wrap the result in a literal
```
boolean result = false;
if (bc instanceof Equals || NullEquals) {
 value = true;
}
else  if (bc instanceof Equals)
...

return new Literal.of(value, bc);
```

costin reviewed (changes requested):
Left a number of comments.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LocalPhysicalPlanOptimizer.java:
```diff
@@ -235,22 +237,44 @@ protected PhysicalPlan rule(FilterExec filterExec, LocalPhysicalOptimizerContext
+                 */
+                if (isPushableAttribute(left, bc, hasIdenticalDelegate) && right.foldable()) {
+                    DataType leftDataType = left.dataType();
+                    if (leftDataType == DataTypes.INTEGER || leftDataType == DataTypes.SHORT || leftDataType == DataTypes.BYTE) {
```
Good point, I'll cover those cases as well. I think I'll also add unsigned longs for peace of mind, even though currently we only support comparisons where both sides are ulongs.

When I first wrote this, I might have assumed that the largest literals we can encounter are `Long.MAX_VALUE` and `Long.MIN_VALUE` - not sure if this was correct then, but it certainly is not now.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LocalPhysicalPlanOptimizer.java:
```diff
@@ -235,22 +237,44 @@ protected PhysicalPlan rule(FilterExec filterExec, LocalPhysicalOptimizerContext
+                Expression left = bc.left();
+                Expression right = bc.right();
+
+                /**
+                 * Because of {@link org.elasticsearch.xpack.esql.type.EsqlDataTypes#widenSmallNumericTypes(DataType)} we should not
+                 * encounter some data types - we include them anyway to be on the safe side.
+                 */
```
I'll clarify the comment (and yes, it should be C-style, oof). What I meant is that the check covers some data types that currently cannot occur on the left hand side (shorts and bytes just become integers in terms of data types due to widening).

I think we should leave the out-of-range check in place; otherwise, this rule will be incorrect unless it's always run after `OutOfRangeBinaryComparison`. This avoids unnecessary dependencies between optimization rules - especially since `OutOfRangeBinaryComparison` happens at a completely different step.

alex-spies replied:
Ended up just removing the changes to the local planner altogether.

alex-spies committed:
Add more test cases

alex-spies committed:
Simplify

alex-spies committed:
Simplify some more

alex-spies committed:
Add scaled_float and unsigned_long to test index

alex-spies commented:
I just realized that optimizing away out-of-range comparisons creates non-equivalent queries because it changes how we handle multi-values.

Consider the following:
```
POST /testidx/_doc"
{"a": [1,2]}

POST /_query
{"query": "from testidx | WHERE a < 1E300 | keep a"}
```

When optimizing away the `a < 1E300` clause (`1E300` is out of range for any integer/long), this will return the single multi-value `[1,2]`.
For an in-range comparison, like `a < 1`, the multi-value `[1,2]` doesn't make it into the result set. So, optimizing away out-of-range comparisons is actually inconsistent with the current semantics of comparisons for multi-values.

In the previous approach (not pushing down the filter into Lucene if it's an out-of-range comparison), the result set does not contain `[1,2]`, which is consistent with the current multi-value semantics in comparisons.

alex-spies committed:
Fix range check for doubles

alex-spies committed:
Perform range check in query translation

Also account for multi-values and widened data types.

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into esql-fix-wrong-type-in-lucene-filter

alex-spies committed:
Add test case with MV literal

alex-spies commented:
Tests still need work, but I pushed an update that improves our query translation rather than performing a logical optimization.

Filters like `WHERE some-int-field < 1E300` now becomes a match all query wrapped in a single value query; `WHERE some-int-field > 1E300` becomes a no-match.

alex-spies committed:
Remove fix for widening of half_float/float

Remove WidenedEsField; this requires amending plan serialization and
probably needs a new transport version, which should be done in a
follow-up.

alex-spies committed:
Handle NaNs correctly

alex-spies committed:
Rework integration test

alex-spies committed:
Ensure we only handle numeric data types

alex-spies commented:
Pushed an update with more complete rest tests - they confirm correct behavior.

I still need to address some remaining comments + add the local physical optimizer tests back in before this is ready for review.

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into esql-fix-wrong-type-in-lucene-filter

alex-spies committed:
Use LoggerMessageFormat

astefan commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/EsqlQueryTranslators.java:
```diff
@@ -0,0 +1,148 @@
+import static org.elasticsearch.xpack.ql.util.NumericUtils.unsignedLongAsNumber;
+
+public class EsqlQueryTranslators {
+    public static final List<ExpressionTranslator<?>> QUERY_TRANSLATORS = Stream.concat(
+        List.of(new TrivialBinaryComparisons()).stream(),
+        ExpressionTranslators.QUERY_TRANSLATORS.stream()
+    ).toList();
```
I think you are complicating this too much creating a list of one, stream it and the stream concatenating it and then creating a list.

```suggestion
    static final List<ExpressionTranslator<?>> QUERY_TRANSLATORS = new ArrayList<>();
    static {
        QUERY_TRANSLATORS.add(new TrivialBinaryComparisons());
        QUERY_TRANSLATORS.addAll(ExpressionTranslators.QUERY_TRANSLATORS);
    }
```

astefan commented on x-pack/plugin/esql/qa/server/src/main/java/org/elasticsearch/xpack/esql/qa/rest/RestEsqlTestCase.java:
```diff
@@ -295,6 +374,105 @@ public void testCSVNoHeaderMode() throws IOException {
+                var query = builder().query(
+                    LoggerMessageFormat.format(null, "from {} | where {} {}", testIndexName(), fieldWithType, trueComparison)
+                );
+                var result = runEsql(query);
```
There are quite a lot of queries this method is testing, around 240.
I would consider introducing some randomization and slim them down a bit.

For example, I would consider the values in `trueForSingleValuesComparisons` in three categories: HUMONGOUS_DOUBLE, INFINITY and NAN. I'd randomly choose one comparison with HUMONGOUS_DOUBLE, one with INFINITY and one with the NAN. With `alwaysFalseComparisons` I'd do something similar, but split the values in four categories: HUMONGOUS_DOUBLE, INFINITY, NAN, to_double and choose randomly one value from each.

alex-spies replied:
I slimmed it down to 66 queries (88 once `half_float` and `float` are enabled).

I wouldn't slim it down any more, the slimmer version already doesn't hit all code paths on every test execution, which could lead to regressions making it into main.

alex-spies committed:
Address Andrei's remarks

alex-spies committed:
Load mapping from file

alex-spies committed:
Update changelog

alex-spies committed:
Fix MV creation in test

alex-spies committed:
Add LocalPhysicalPlanOptimizerTests back in

alex-spies committed:
Remove leftover from Analyzer

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into esql-fix-wrong-type-in-lucene-filter

alex-spies committed:
Spotless

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

alex-spies commented:
Tests done, this should be ready for another round of reviews now.

alex-spies committed:
Make our linter happy

costin commented on x-pack/plugin/esql/qa/server/src/main/java/org/elasticsearch/xpack/esql/qa/rest/RestEsqlTestCase.java:
```diff
@@ -71,6 +73,30 @@ public abstract class RestEsqlTestCase extends ESRestTestCase {
+    private static final String MAPPING_ALL_TYPES;
+
+    static {
+        try {
+            InputStream mappingPropertiesStream = RestEsqlTestCase.class.getResourceAsStream("/mapping-all-types.json");
+            String properties = new String(mappingPropertiesStream.readAllBytes(), StandardCharsets.UTF_8);
+            MAPPING_ALL_TYPES = "{\"mappings\": " + properties + "}";
+        } catch (IOException ex) {
+            throw new RuntimeException(ex);
+        }
```
You need to close the stream in case of errors - hence use the [try-with-resource](https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html) pattern.
Also extract the code into a static method and use that for field initialization:

```java
@BeforeClass
public static void initClass() {
   MAPPING_ALL_TYPES = loadResourceAsString("/mapping-all-types.json");
}

public static String loadResourceAsString(String name) throws Exception {
try (var url = RestEsqlTestCase.class.getResource(name)) {
    return Files.readString(Paths.get(url.toURI()), StandardCharsets.UTF_8));
}
```
There's probably an utility somewhere in ES that does something similar.

alex-spies replied:
:facepalm: of course, thank you.

I couldn't find something that _just_ loads a file as text; most utilities seem to directly parse that (e.g. `EsqlTestUtils.loadmapping` and `TypesTests.loadMapping`) or not expose that (e.g. `GeometrySimplifierTests.loadJsonFile`) :/

costin commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LocalPhysicalPlanOptimizerTests.java:
```diff
@@ -396,6 +384,112 @@ public void testIsNullPushdownFilter() {
+                String comparison = testCase.fieldName + falsePredicate;
+                var query = "from test | where " + comparison;
+                Source expectedSource = new Source(1, 18, comparison);
+                String expectedLuceneQuery = LoggerMessageFormat.format(null, """
+                    {
+                      "esql_single_value" : {
+                        "field" : "{}",
+                        "next" : {
+                          "bool" : {
+                            "must_not" : [
+                              {
+                                "match_all" : {
+                                  "boost" : 1.0
+                                }
+                              }
+                            ],
+                            "boost" : 1.0
+                          }
+                        },
+                        "source" : "{}"
+                      }
```
String comparisons are finicky - next time walk the object tree.
check the overall query is QSV, then its wrapped query and that is either is match_all or a negated one.

costin reviewed (commented):
Thanks - this looks good.
Left some comments around style and unnecessary delegation.

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into esql-fix-wrong-type-in-lucene-filter

alex-spies committed:
Move new tranlator into EsqlQueryTranslators

alex-spies committed:
Revert QL ExpressionTranslators

alex-spies committed:
Explicitly pass null in LoggerMessageFormat.format

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/EsqlTranslatorHandler.java:
```diff
@@ -27,7 +27,7 @@
     @Override
     public Query asQuery(Expression e) {
-        return ExpressionTranslators.toQuery(e, this);
+        return ExpressionTranslators.toQuery(e, this, EsqlQueryTranslators.QUERY_TRANSLATORS);
```
I see that in the meantime, Luigi's `=~` PR added ESQL-specific query translators - I'll align with the now existing code and will move my code next to his.

alex-spies committed:
Walk object tree instead of string comparisons

alex-spies commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LocalPhysicalPlanOptimizerTests.java:
```diff
@@ -113,24 +111,9 @@ public LocalPhysicalPlanOptimizerTests(String name, EsqlConfiguration config) {
     @Before
     public void init() {
         parser = new EsqlParser();
-
-        mapping = loadMapping("mapping-basic.json");
-        allFieldRowSize = mapping.values()
-            .stream()
-            .mapToInt(
-                f -> (EstimatesRowSize.estimateSize(EsqlDataTypes.widenSmallNumericTypes(f.getDataType())) + f.getProperties()
-                    .values()
-                    .stream()
-                    // check one more level since the mapping contains TEXT fields with KEYWORD multi-fields
-                    .mapToInt(x -> EstimatesRowSize.estimateSize(EsqlDataTypes.widenSmallNumericTypes(x.getDataType())))
-                    .sum())
-            )
-            .sum();
-        EsIndex test = new EsIndex("test", mapping);
```
This was unused and looked like a leftover, so I removed it. (Wanted to clean this up as I'm already touching directly adjacent code.)

Should this stay here, after all?

costin replied:
No, deleting code is good but it was unclear since it was unrelated to this PR.

alex-spies committed:
Use stream with try-with-resources

astefan commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LocalPhysicalPlanOptimizerTests.java:
```diff
@@ -422,6 +410,108 @@ public void testIsNullPushdownFilter() {
+     *       \_EsQueryExec[test], query[{"esql_single_value":{"field":"byte","next":{"match_all":{"boost":1.0}},...}}]
+     */
+    private EsQueryExec doTestOutOfRangeFilterPushdown(String query) {
+        var analyzer = makeAnalyzer("mapping-all-types.json", new EnrichResolution());
```
Here you are creating a new Analyzer for every method call and this is actually not needed.

astefan reviewed (approved):
I've left few more clean up comments, otherwise LGTM

costin commented on x-pack/plugin/esql/qa/server/src/main/java/org/elasticsearch/xpack/esql/qa/rest/RestEsqlTestCase.java:
```diff
@@ -295,6 +320,81 @@ public void testCSVNoHeaderMode() throws IOException {
+
+                var values = (ArrayList) result.get("values");
+                assertThat(
+                    LoggerMessageFormat.format(null, "Comparison [{}] should return all rows with single values.", comparison),
```
Use static imports.

costin commented on x-pack/plugin/esql/qa/server/src/main/java/org/elasticsearch/xpack/esql/qa/rest/RestEsqlTestCase.java:
```diff
@@ -295,6 +320,81 @@ public void testCSVNoHeaderMode() throws IOException {
+                );
+                var result = runEsql(query, expectedWarnings, mode);
+
+                var values = (ArrayList) result.get("values");
```
Instead of doing casting import the as method from EsqTestUtils:
`var values = as(result.get("values"), List.class)`

alex-spies replied:
Will do! EsqlUtils requires to re-export esql in Gradle, otherwise the class won't load - but that's a small change.

costin commented on x-pack/plugin/esql/qa/server/src/main/java/org/elasticsearch/xpack/esql/qa/rest/RestEsqlTestCase.java:
```diff
@@ -295,6 +320,81 @@ public void testCSVNoHeaderMode() throws IOException {
+            "scaled_float"
+        );
+
+        String lessOrLessEqual = randomBoolean() ? " < " : " <= ";
+        String largerOrLargerEqual = randomBoolean() ? " > " : " >= ";
+        String inEqualPlusMinus = randomBoolean() ? " != " : " != -";
+        String equalPlusMinus = randomBoolean() ? " == " : " == -";
+        // TODO: once we do not support infinity and NaN anymore, remove INFINITY/NAN cases.
+        // https://github.com/elastic/elasticsearch/issues/98698#issuecomment-1847423390
+        String humongousPositiveLiteral = randomBoolean() ? HUMONGOUS_DOUBLE : INFINITY;
+        String nanOrNull = randomBoolean() ? NAN : "to_double(null)";
```
A shorter and better alternative to the above is to use `randomFrom` which works with 2+ elements and has the same effect.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/EsqlTranslatorHandler.java:
```diff
@@ -22,23 +22,39 @@
+
 public final class EsqlTranslatorHandler extends QlTranslatorHandler {
 
     public static final List<ExpressionTranslator<?>> QUERY_TRANSLATORS = List.of(
```
This PR mixes ExpressionTranslators (which is a list of translators) with TranslatorHandler which is the pluggable component - namely the only thing that needs to be implemented.

Extract the list and the `TrivialBinaryComparisons` to a separate `EsqlExpressionTranslators` and have the `EsqlTranslatorHandler` delegate to it.

alex-spies replied:
I agree - I initially had the new translator in a dedicated `EsqlExpressionTranslators`, but then moved it here to align that was added in the meantime.

More specifically, `EsqlTranslatorHandler` already contains `EqualsIgnoreCaseTranslator`. I can move _both_ into a dedicated `EsqlExpressionTranslators`, but I'd do so in a separate PR to avoid touching the `EqualsIgnoreCaseTranslator`.

Is that good with you?

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/EsqlTranslatorHandler.java:
```diff
@@ -22,23 +22,39 @@
     public static final List<ExpressionTranslator<?>> QUERY_TRANSLATORS = List.of(
         new EqualsIgnoreCaseTranslator(),
+        new TrivialBinaryComparisons(),
         new ExpressionTranslators.BinaryComparisons(),
```
Is the BinaryComparison actually used? Isn't TrivialBinaryComparison picked up instead? Why not delegate from that to BinaryComparison?

alex-spies replied:
If `TrivialBinaryComparison` doesn't apply, it returns null, and is thus skipped. Then `BinaryComparison` applies; I 

Equivalently, I can replace both by a single translator that delegates to `BinaryComparison` if out-of-range comparisons don't apply. Should I go for it?

alex-spies replied:
(Also, lots of local physical optimizer tests fail without `BinaryComparison`.)

costin commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LocalPhysicalPlanOptimizerTests.java:
```diff
@@ -422,6 +410,108 @@ public void testIsNullPushdownFilter() {
+    private record OutOfRangeTestCase(String fieldName, String tooLow, String tooHigh) {};
+
+    public void testOutOfRangeFilterPushdown() {
+        String largerThanInteger = ((Long) randomLongBetween(Integer.MAX_VALUE + 1L, Long.MAX_VALUE)).toString();
+        String smallerThanInteger = ((Long) randomLongBetween(Long.MIN_VALUE, Integer.MIN_VALUE - 1L)).toString();
```
Use `String.valueOf(long)` instead of "" + randomLong...

costin commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LocalPhysicalPlanOptimizerTests.java:
```diff
@@ -422,6 +410,108 @@ public void testIsNullPushdownFilter() {
+        // These values are already out of bounds for longs due to rounding errors.
+        double longLowerBoundExclusive = (double) Long.MIN_VALUE;
+        double longUpperBoundExclusive = (double) Long.MAX_VALUE;
+        String largerThanLong = ((Double) randomDoubleBetween(longUpperBoundExclusive, Double.MAX_VALUE, true)).toString();
+        String smallerThanLong = ((Double) randomDoubleBetween(-Double.MAX_VALUE, longLowerBoundExclusive, true)).toString();
```
Same here.

costin commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LocalPhysicalPlanOptimizerTests.java:
```diff
@@ -422,6 +410,108 @@ public void testIsNullPushdownFilter() {
+
+        for (OutOfRangeTestCase testCase : cases) {
+            List<String> trueForSingleValuesPredicates = List.of(
+                "<" + testCase.tooHigh,
+                "<=" + testCase.tooHigh,
+                ">" + testCase.tooLow,
+                ">=" + testCase.tooLow,
+                "!=" + testCase.tooHigh,
+                "!=" + testCase.tooLow,
+                "!= 0.0/0.0"
```
Define constants for the strings - LT, LTE...

costin reviewed (commented):
Need a bit more polishing when it comes to delegating.

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into esql-fix-wrong-type-in-lucene-filter

alex-spies committed:
Clean RestEsqlTestCase

alex-spies committed:
Clean LocalPhysicalPlanOptimizerTests

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/EsqlTranslatorHandler.java:
```diff
@@ -124,4 +140,110 @@ static Query translate(InsensitiveEquals bc) {
+        @Override
+        protected Query asQuery(BinaryComparison bc, TranslatorHandler handler) {
+            ExpressionTranslators.BinaryComparisons.checkBinaryComparison(bc);
+            Query translated = translate(bc, handler);
+            return translated == null ? null : handler.wrapFunctionQuery(bc, bc.left(), () -> translated);
```
This leads to NPEs because we use the overridden `wrapFunctionQuery`. I could make it deal with nulls, but that could accidentally mask unintended null queries, no?

astefan replied:
I missed that it's using the current handler function. All good.

alex-spies committed:
Clean EsqlTranslatorHandler

astefan commented on x-pack/plugin/esql/qa/server/single-node/src/javaRestTest/java/org/elasticsearch/xpack/esql/qa/single_node/RestEsqlIT.java:
```diff
@@ -2,7 +2,7 @@
  * or more contributor license agreements. Licensed under the Elastic License
  * 2.0; you may not use this file except in compliance with the Elastic License
- * 2.0.
+ * 2.0.LoggerM
```
Leftover?

alex-spies committed:
Remove leftover

alex-spies commented:
Thanks for reviews, Andrei and Costin!

alex-spies commented:
And thanks for the super fast review of the `build.gradle` change, Rene!
